src/tcache.c

```diff
@@ -703,36 +678,41 @@ tcache_init(tsd_t *tsd, tcache_slow_t *tcache_slow, tcache_t *tcache,
 		    &cur_offset);
 	}
 	/*
-	 * For small size classes beyond tcache_max(i.e.
-	 * tcache_nhbins< NBINS), their cache bins are initialized to a state
-	 * to safely and efficiently fail all fastpath alloc / free, so that
-	 * no additional check around tcache_nhbins is needed on fastpath.
+	 * Initialize all disabled bins to a state that can safely and
+	 * efficiently fail all fastpath alloc / free, so that no additional
+	 * check around tcache_nbins is needed on fastpath.  Yet we still
+	 * store the ncached_max in the bin_info for future usage.
 	 */
-	for (unsigned i = tcache_nhbins; i < SC_NBINS; i++) {
-		/* Disabled small bins. */
+	const void *fake_stack =
+	    cache_bin_disabled_bin_stack();
+	size_t fake_offset = 0;
+	cache_bin_info_t fake_info;
+	cache_bin_info_init(&fake_info, 0);
+
+	for (unsigned i = tcache_nbins; i < TCACHE_NBINS_MAX; i++) {
 		cache_bin_t *cache_bin = &tcache->bins[i];
-		void *fake_stack = mem;
-		size_t fake_offset = 0;
+		cache_bin_init(cache_bin, &fake_info,
```

Comment:
I think for better readability, we can do `cache_bin_init_disabled(cache_bin, tcache_bin_info[i].ncached_max)` so that we could hide the fake initialization details there? Going to sort the fast path since now we have large bins also fake initialized. Maybe a good idea to make this change there as well?

Reply:
Sounds good. Please feel free to change this part. Certainly doesn't hurt to make it clean sooner.

Reply:
Done.